pyro/infer/mcmc/hmc.py

```diff
@@ -194,34 +187,38 @@ def _find_reasonable_step_size(self, z):
         while direction_new == direction:
             t += 1
             step_size = step_size_scale * step_size
-            r, _ = self._sample_r(name="r_presample_{}".format(t))
-            energy_current = self._kinetic_energy(r) + potential_energy
+            r, r_unscaled = self._sample_r(name="r_presample_{}".format(t))
+            energy_current = self._kinetic_energy(r_unscaled) + potential_energy
             z_new, r_new, z_grads_new, potential_energy_new = velocity_verlet(
-                z, r, self.potential_fn, self.inverse_mass_matrix, step_size)
-            energy_new = self._kinetic_energy(r_new) + potential_energy_new
+                z, r, self.potential_fn, self.mass_matrix_adapter.kinetic_grad, step_size)
+            r_new_unscaled = self.mass_matrix_adapter.unscale(r_new)
+            energy_new = self._kinetic_energy(r_new_unscaled) + potential_energy_new
             delta_energy = energy_new - energy_current
             direction_new = 1 if self._direction_threshold < -delta_energy else -1
         return step_size
 
     def _sample_r(self, name):
-        r = {}
-        r_flat_dict = {}
-        for site_names in self.inverse_mass_matrix:
-            r_flat = pyro.sample(
-                "{}_{}".format(name, site_names), self._adapter.r_dist[site_names])
-            pos = 0
-            for name in site_names:
-                param = self.initial_params[name]
-                next_pos = pos + param.numel()
-                r[name] = r_flat[pos:next_pos].reshape(param.shape)
-                pos = next_pos
-            assert pos == r_flat.size(0)
-            r_flat_dict[site_names] = r_flat
-        return r, r_flat_dict
+        r_unscaled = {}
+        for site_names, value in self.inverse_mass_matrix.items():
+            size = value.size(0)
+            r_unscaled[site_names] = pyro.sample(
+                "{}_{}".format(name, site_names),
+                NonreparameterizedNormal(value.new_zeros(size), value.new_ones(size)))
```

Comment:
Could you explain what's going on here, and maybe add a comment explaining the use of `NonreparameterizedNormal` rather than `Normal`?

Reply:
There is nothing special here. It is a bit faster than `Normal` I guess (I'll add a comment). I think `torch.normal(loc, scale)` is faster than `loc  + scale * torch.randn(shape)`.